geotiepoints/simple_modis_interpolator.py

```diff
@@ -0,0 +1,255 @@
+        return lon_data, lat_data
+
+    new_row_chunks = (row_chunks[0] // ROWS_PER_SCAN) * ROWS_PER_SCAN
+    lon_data = lon_data.rechunk((new_row_chunks, -1))
+    lat_data = lat_data.rechunk((new_row_chunks, -1))
+    return lon_data, lat_data
+
+
+def _map_blocks_handler(func):
+    def _map_blocks_wrapper(lon_array, lat_array, res_factor):
+        lons, lats = func(lon_array, lat_array, res_factor=res_factor)
+        return np.concatenate((lons[np.newaxis], lats[np.newaxis]), axis=0)
+    return _map_blocks_wrapper
+
+
+@scanline_mapblocks
+def interpolate_geolocation_cartesian(lon_array, lat_array, res_factor=4):
+    """Interpolate MODIS navigation from 1000m resolution to 250m.
+
+    Python rewrite of the IDL function ``MODIS_GEO_INTERP_250`` but converts to cartesian (X, Y, Z) coordinates
+    first to avoid problems with the anti-meridian/poles.
+
+    Arguments:
+        lon_array: Longitude data as a 2D numpy, dask, or xarray DataArray object.
+            The input data is expected to represent 1000m geolocation.
+        lat_array: Latitude data as a 2D numpy, dask, or xarray DataArray object.
+            The input data is expected to represent 1000m geolocation.
+        res_factor (int): Expansion factor for the function. Should be 2 for
+            500m output or 4 for 250m output.
+
+    Returns:
+        A two-element tuple (lon, lat).
+
+    """
+    num_rows, num_cols = lon_array.shape
+    num_scans = int(num_rows / ROWS_PER_SCAN)
+    x_in, y_in, z_in = lonlat2xyz(lon_array, lat_array)
+
+    # Create an array of indexes that we want our result to have
+    x = np.arange(res_factor * num_cols, dtype=np.float32) * (1. / res_factor)
+    # 0.375 for 250m, 0.25 for 500m
+    y = np.arange(res_factor * ROWS_PER_SCAN, dtype=np.float32) * \
+        (1. / res_factor) - (res_factor * (1. / 16) + (1. / 8))
+    x, y = np.meshgrid(x, y)
+    coordinates = np.array([y, x])  # Used by map_coordinates, major optimization
+
+    new_x = np.empty((num_rows * res_factor, num_cols * res_factor), dtype=lon_array.dtype)
+    new_y = new_x.copy()
+    new_z = new_x.copy()
+    nav_arrays = [(x_in, new_x), (y_in, new_y), (z_in, new_z)]
+
+    # Interpolate each scan, one at a time, otherwise the math doesn't work well
+    for scan_idx in range(num_scans):
+        # Calculate indexes
+        j0 = ROWS_PER_SCAN * scan_idx
+        j1 = j0 + ROWS_PER_SCAN
+        k0 = ROWS_PER_SCAN * res_factor * scan_idx
+        k1 = k0 + ROWS_PER_SCAN * res_factor
+
+        for nav_array, result_array in nav_arrays:
+            # Use bilinear interpolation for all 250 meter pixels
+            map_coordinates(nav_array[j0:j1, :], coordinates, output=result_array[k0:k1, :], order=1, mode='nearest')
+
+            if res_factor == 4:
+                # Use linear extrapolation for the first two 250 meter pixels along track
+                m = (result_array[k0 + 5, :] - result_array[k0 + 2, :]) / (y[5, 0] - y[2, 0])
```

Comment:
These for lines are duplicated, could they be factorized?

Reply:
They aren't exactly repeated. They have different indexes. Given that there isn't a difference in number of lines (unless I'm reading it wrong) I think I'd prefer to keep it this way since it is more clear, at least to me, what the logic is trying to do. I'd be worried that putting things in a separate function would make it harder to parse this out.

Reply:
The computation of m and b (slope and offset) is identical with an offset of 32, right? I think they should be factorized. Maybe even with the 500m, just passing the indices.